File: playground/src/compiler/parser.ts
import type { ArchFlowAst, Component, Diagnostic, FlowNode, Workflow } from "./types";

const ATTR_RE = /^\.(?<key>[^:]+):\s*(?<value>.*)$/;
const CHANNEL_RE = /^@(?<source>[A-Za-z_]\w*)\s*=>\s*\[(?<via>[^\]]+)\]\s*=>\s*@(?<target>[A-Za-z_]\w*)$/;
const COMPONENT_RE = /^@(?<name>[A-Za-z_]\w*):?$/;
const WORKFLOW_RE = /^\$(?<name>[A-Za-z_]\w*):$/;
const EXCEPTION_RE = /^!\s*(?<name>[A-Za-z_]\w*)\s*(?<tail>>>.*)?$/;

export function parseArchFlow(text: string): ArchFlowAst {
  const ast: ArchFlowAst = {
    version: "0.1",
    attributes: {},
    components: {},
    channels: [],
    diagnostics: [],
  };

  let currentComponent: string | undefined;
  let currentWorkflow: Workflow | undefined;

  text.split(/\r?\n/).forEach((rawLine, index) => {
    const lineNo = index + 1;
    const stripped = rawLine.trim();
    if (!stripped || stripped.startsWith("#")) {
      return;
    }

    const indent = rawLine.length - rawLine.trimStart().length;
    if (rawLine.slice(0, indent).includes("\t")) {
      ast.diagnostics.push(diagnostic("error", "Tabs are not supported for indentation", lineNo));
      return;
    }
    if (indent % 2 !== 0) {
      ast.diagnostics.push(diagnostic("error", "Indentation must use multiples of two spaces", lineNo));
      return;
    }

    if (indent === 0) {
      currentComponent = undefined;
      currentWorkflow = undefined;
      currentComponent = parseRootLine(ast, stripped, lineNo);
      return;
    }

    if (indent === 2) {
      currentWorkflow = undefined;
      if (!currentComponent) {
        ast.diagnostics.push(diagnostic("error", "Module-level line appears outside a component block", lineNo));
        return;
      }
      parseComponentLine(ast, currentComponent, stripped, lineNo);
      const component = ast.components[currentComponent];
      currentWorkflow = component.workflows.at(-1);
      return;
    }

    if (indent === 4) {
      if (!currentComponent || !currentWorkflow) {
        ast.diagnostics.push(diagnostic("error", "Workflow line appears outside a workflow block", lineNo));
        return;
      }
      parseWorkflowLine(ast, currentWorkflow, stripped, lineNo);
      return;
    }

    ast.diagnostics.push(diagnostic("error", "Only root, module, and workflow indentation levels are supported", lineNo));
  });

  bindChannelSchemas(ast);
  return ast;
}

function parseRootLine(ast: ArchFlowAst, line: string, lineNo: number): string | undefined {
  const attr = ATTR_RE.exec(line);
  if (attr?.groups) {
    ast.attributes[attr.groups.key.trim()] = attr.groups.value.trim();
    return undefined;
  }

  const channel = CHANNEL_RE.exec(line);
  if (channel?.groups) {
    const source = channel.groups.source;
    const target = channel.groups.target;
    ensureComponent(ast, source, false, lineNo);
    ensureComponent(ast, target, false, lineNo);
    ast.channels.push({
      source,
      target,
      via: channel.groups.via.trim(),
      line: lineNo,
    });
    return undefined;
  }

  const component = COMPONENT_RE.exec(line);
  if (component?.groups) {
    const name = component.groups.name;
    ensureComponent(ast, name, true, lineNo);
    return name;
  }

  ast.diagnostics.push(diagnostic("error", `Unsupported root syntax: ${line}`, lineNo));
  return undefined;
}

function parseComponentLine(ast: ArchFlowAst, componentName: string, line: string, lineNo: number): void {
  const component = ast.components[componentName];
  const attr = ATTR_RE.exec(line);
  if (attr?.groups) {
    component.attributes[attr.groups.key.trim()] = attr.groups.value.trim();
    return;
  }

  const workflow = WORKFLOW_RE.exec(line);
  if (workflow?.groups) {
    component.workflows.push({
      name: workflow.groups.name,
      sequence: [],
      exceptions: [],
      line: lineNo,
    });
    return;
  }

  ast.diagnostics.push(diagnostic("error", `Unsupported component syntax: ${line}`, lineNo));
}

function parseWorkflowLine(ast: ArchFlowAst, workflow: Workflow, line: string, lineNo: number): void {
  const exception = EXCEPTION_RE.exec(line);
  if (exception?.groups) {
    const tail = exception.groups.tail ?? "";
    workflow.exceptions.push({
      name: exception.groups.name,
      sequence: parseSequence(tail.replace(/^>>/, "").trim()),
      line: lineNo,
    });
    return;
  }

  const sequence = parseSequence(line);
  if (sequence.length > 0) {
    workflow.sequence.push(...sequence);
    return;
  }

  ast.diagnostics.push(diagnostic("error", `Unsupported workflow syntax: ${line}`, lineNo));
}

function parseSequence(value: string): FlowNode[] {
  return value
    .split(">>")
    .map((part) => part.trim())
    .filter(Boolean)
    .map(parseFlowNode);
}

function parseFlowNode(value: string): FlowNode {
  if (value.startsWith("[") && value.endsWith("]")) {
    return { kind: "action", value: value.slice(1, -1).trim() };
  }
  if (value.startsWith("(") && value.endsWith(")")) {
    return { kind: "state", value: value.slice(1, -1).trim() };
  }
  if (value.startsWith("?")) {
    return { kind: "condition", value: value.slice(1).trim() };
  }
  if (value.startsWith("~")) {
    return { kind: "inference", value: value.slice(1).trim() };
  }
  return { kind: "raw", value };
}

function bindChannelSchemas(ast: ArchFlowAst): void {
  ast.channels.forEach((channel) => {
    channel.schema = ast.attributes[`Schema.${channel.via}`];
  });
}

function ensureComponent(ast: ArchFlowAst, name: string, explicit: boolean, line?: number): Component {
  const existing = ast.components[name];
  if (existing) {
    if (explicit) {
      existing.explicit = true;
      existing.line = line;
    }
    return existing;
  }

  const component: Component = {
    name,
    explicit,
    attributes: {},
    workflows: [],
    line,
  };
  ast.components[name] = component;
  return component;
}

function diagnostic(severity: Diagnostic["severity"], message: string, line?: number): Diagnostic {
  return { severity, message, line };
}

File: compiler/archflow/parser.py
from __future__ import annotations

import re
from pathlib import Path

from .model import ArchFlowAst, Channel, Diagnostic, ExceptionFlow, FlowNode, Workflow

ATTR_RE = re.compile(r"^\.(?P<key>[^:]+):\s*(?P<value>.*)$")
CHANNEL_RE = re.compile(r"^@(?P<source>[A-Za-z_][\w]*)\s*=>\s*\[(?P<via>[^\]]+)\]\s*=>\s*@(?P<target>[A-Za-z_][\w]*)$")
COMPONENT_RE = re.compile(r"^@(?P<name>[A-Za-z_][\w]*):?$")
WORKFLOW_RE = re.compile(r"^\$(?P<name>[A-Za-z_][\w]*):$")
EXCEPTION_RE = re.compile(r"^!\s*(?P<name>[A-Za-z_][\w]*)\s*(?P<tail>>>.*)?$")


class ParseError(ValueError):
    """Raised when a source file cannot be read before parsing begins."""


def parse_file(path: str | Path) -> ArchFlowAst:
    source_path = Path(path)
    try:
        return parse_text(source_path.read_text(encoding="utf-8"))
    except OSError as exc:
        raise ParseError(f"Unable to read {source_path}: {exc}") from exc


def parse_text(text: str) -> ArchFlowAst:
    ast = ArchFlowAst()
    current_component: str | None = None
    current_workflow: Workflow | None = None

    for line_no, raw_line in enumerate(text.splitlines(), start=1):
        stripped = raw_line.strip()
        if not stripped or stripped.startswith("#"):
            continue

        indent = len(raw_line) - len(raw_line.lstrip(" "))
        if "\t" in raw_line[:indent]:
            ast.diagnostics.append(Diagnostic("error", "Tabs are not supported for indentation", line_no))
            continue
        if indent % 2 != 0:
            ast.diagnostics.append(Diagnostic("error", "Indentation must use multiples of two spaces", line_no))
            continue

        if indent == 0:
            current_component = None
            current_workflow = None
            current_component = _parse_root_line(ast, stripped, line_no)
        elif indent == 2:
            current_workflow = None
            if current_component is None:
                ast.diagnostics.append(Diagnostic("error", "Module-level line appears outside a component block", line_no))
                continue
            _parse_component_line(ast, current_component, stripped, line_no)
            component = ast.components[current_component]
            current_workflow = component.workflows[-1] if component.workflows else None
        elif indent == 4:
            if current_component is None or current_workflow is None:
                ast.diagnostics.append(Diagnostic("error", "Workflow line appears outside a workflow block", line_no))
                continue
            _parse_workflow_line(ast, current_workflow, stripped, line_no)
        else:
            ast.diagnostics.append(Diagnostic("error", "Only root, module, and workflow indentation levels are supported", line_no))

    _bind_channel_schemas(ast)
    return ast


def _parse_root_line(ast: ArchFlowAst, line: str, line_no: int) -> str | None:
    attr = ATTR_RE.match(line)
    if attr:
        ast.attributes[attr.group("key").strip()] = attr.group("value").strip()
        return None

    channel = CHANNEL_RE.match(line)
    if channel:
        source = channel.group("source")
        target = channel.group("target")
        ast.ensure_component(source, explicit=False, line=line_no)
        ast.ensure_component(target, explicit=False, line=line_no)
        ast.channels.append(Channel(source=source, via=channel.group("via").strip(), target=target, line=line_no))
        return None

    component = COMPONENT_RE.match(line)
    if component:
        ast.ensure_component(component.group("name"), explicit=True, line=line_no)
        return component.group("name")

    ast.diagnostics.append(Diagnostic("error", f"Unsupported root syntax: {line}", line_no))
    return None


def _parse_component_line(ast: ArchFlowAst, component_name: str, line: str, line_no: int) -> None:
    component = ast.components[component_name]
    attr = ATTR_RE.match(line)
    if attr:
        component.attributes[attr.group("key").strip()] = attr.group("value").strip()
        return

    workflow = WORKFLOW_RE.match(line)
    if workflow:
        component.workflows.append(Workflow(name=workflow.group("name"), line=line_no))
        return

    ast.diagnostics.append(Diagnostic("error", f"Unsupported component syntax: {line}", line_no))


def _parse_workflow_line(ast: ArchFlowAst, workflow: Workflow, line: str, line_no: int) -> None:
    exception = EXCEPTION_RE.match(line)
    if exception:
        tail = exception.group("tail") or ""
        workflow.exceptions.append(
            ExceptionFlow(
                name=exception.group("name"),
                sequence=_parse_sequence(tail.removeprefix(">>").strip()),
                line=line_no,
            )
        )
        return

    sequence = _parse_sequence(line)
    if sequence:
        workflow.sequence.extend(sequence)
        return

    ast.diagnostics.append(Diagnostic("error", f"Unsupported workflow syntax: {line}", line_no))


def _parse_sequence(value: str) -> list[FlowNode]:
    nodes: list[FlowNode] = []
    for part in [piece.strip() for piece in value.split(">>") if piece.strip()]:
        nodes.append(_parse_flow_node(part))
    return nodes


def _parse_flow_node(value: str) -> FlowNode:
    if value.startswith("[") and value.endswith("]"):
        return FlowNode(kind="action", value=value[1:-1].strip())
    if value.startswith("(") and value.endswith(")"):
        return FlowNode(kind="state", value=value[1:-1].strip())
    if value.startswith("?"):
        return FlowNode(kind="condition", value=value[1:].strip())
    if value.startswith("~"):
        return FlowNode(kind="inference", value=value[1:].strip())
    return FlowNode(kind="raw", value=value)


def _bind_channel_schemas(ast: ArchFlowAst) -> None:
    for channel in ast.channels:
        schema_key = f"Schema.{channel.via}"
        channel.schema = ast.attributes.get(schema_key)
